File: leetcode-04/solutions/python3/0076.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if not s or not t:
            return ""

        dict = collections.Counter(t)

        l = 0
        r = 0
        bestLeft = 0
        bestRight = 0
        required = len(dict)
        windowLength = len(s) + 1

        for r in range(len(s)):
            if s[r] in dict:
                dict[s[r]] -= 1
                if dict[s[r]] == 0:
                    required -= 1

            while required == 0 and l <= r:
                if r - l + 1 < windowLength:
                    windowLength = r - l + 1
                    bestLeft = l
                    bestRight = r

                if s[l] in dict:
                    dict[s[l]] += 1
                    if dict[s[l]] > 0:
                        required += 1

                l += 1

        return "" if windowLength == len(s) + 1 else s[bestLeft: bestRight + 1]

File: leetcode-04/solutions/cpp/0076.cpp
class Solution {
 public:
  string minWindow(string s, string t) {
    if (s.empty() || t.empty()) return "";

    unordered_map<char, int> map;
    for (char c : t) map[c]++;

    int l = 0;
    int r = 0;
    int bestLeft = 0;
    int bestRight = 0;
    int required = map.size();
    int windowLength = s.length() + 1;

    for (int r = 0; r < s.length(); r++) {
      if (map.count(s[r]))
        if (--map[s[r]] == 0) required--;

      while (required == 0 && l <= r) {
        if (r - l + 1 < windowLength) {
          windowLength = r - l + 1;
          bestLeft = l;
          bestRight = r;
        }

        if (map.count(s[l]))
          if (++map[s[l]] > 0) required++;

        l++;
      }
    }

    return windowLength == s.length() + 1 ? ""
                                          : s.substr(bestLeft, windowLength);
  }
};